.ci/breaking-change-detector/comparison.go

```diff
@@ -0,0 +1,128 @@
+			for _, resourceName := range violatingResources {
+				newMessage := rule.Message(*providerVersion, resourceName)
+				messages = append(messages, newMessage)
+			}
+		}
+
+	}
+
+	for resourceName, resource := range new {
+		oldResource, ok := old[resourceName]
+		if ok {
+			newMessages := compareResourceSchema(resourceName, oldResource.Schema, resource.Schema)
+			messages = append(messages, newMessages...)
+		}
+	}
+
+	return messages
+}
+
+func compareResourceSchema(resourceName string, old, new map[string]*schema.Schema) []string {
+	messages := []string{}
+	oldCompressed := flattenSchema(old)
+	newCompressed := flattenSchema(new)
+
+	for _, rule := range rules.ResourceSchemaRules {
+		violatingFields := rule.IsRuleBreak(old, new)
+		if len(violatingFields) > 0 {
+			for _, fieldName := range violatingFields {
+				newMessage := rule.Message(*providerVersion, resourceName, fieldName)
+				messages = append(messages, newMessage)
+			}
+		}
+	}
+
+	for fieldName, field := range newCompressed {
+		oldField, ok := oldCompressed[fieldName]
+		if ok {
+			newMessages := compareField(resourceName, fieldName, oldField, field)
+			messages = append(messages, newMessages...)
+		}
+	}
+
+	return messages
+}
+
+func compareField(resourceName, fieldName string, old, new *schema.Schema) []string {
+	messages := []string{}
+	rules := rules.FieldRules
+
+	for _, rule := range rules {
+		isBreakage := rule.IsRuleBreak(old, new)
+		if isBreakage {
+			newMessage := rule.Message(*providerVersion, resourceName, fieldName)
+			messages = append(messages, newMessage)
+		}
+	}
+
+	return messages
+}
+
+func flattenSchema(schemaObj map[string]*schema.Schema) map[string]*schema.Schema {
+	return flattenSchemaRecursive(nil, schemaObj)
+}
+
+func flattenSchemaRecursive(parentLineage []string, schemaObj map[string]*schema.Schema) map[string]*schema.Schema {
+	compressed := make(map[string]*schema.Schema)
+
+	// prepare prefix to bring nested entries up
+	parentPrefix := strings.Join(parentLineage, ".")
+	if len(parentPrefix) > 0 {
+		parentPrefix += "."
+	}
+
+	// add entry to output and call
+	// flattenSchemaRecursive for any children
+	for fieldName, field := range schemaObj {
+		compressed[parentPrefix+fieldName] = field
+		casted, typeConverted := field.Elem.(*schema.Resource)
+		if field.Elem != nil && typeConverted {
+			newLineage := append([]string{}, parentLineage...)
+			newLineage = append(newLineage, fieldName)
+			compressedChild := flattenSchemaRecursive(newLineage, casted.Schema)
+			compressed = mergeSchemaMaps(compressed, compressedChild)
+		}
+	}
+
+	return compressed
+}
+
+func mergeSchemaMaps(map1, map2 map[string]*schema.Schema) map[string]*schema.Schema {
+	merged := make(map[string]*schema.Schema)
+	for key, value := range map1 {
+		merged[key] = value
+	}
+
+	for key, value := range map2 {
+		if _, alreadyExists := merged[key]; alreadyExists {
+			panic("when trying to merge maps key " + key + " was found in both maps")
```

Comment:
What does this indicate and how would someone seeing this try to resolve it?

Reply:
Also, should this be glog.Exit or Error?

Reply:
This error should never happen in practice. It means something has gone wrong in merging the nested fields down since we are always merging children up with prefixes attached.

Reply:
Changed to errorf and added more information